### ISSUE TITLE
OnRowMouseLeave executed with wrong index

### ISSUE THREAD

**Issue by SirGandal** [schrodinger/fixed-data-table-2#247]
_OnRowMouseLeave_ executed with the wrong index after having hovered on a specific row and having scrolled without moving the mouse and in such a way that the row we initially hovered on is recycled. 

## Expected Behavior
_onRowMouseLeave_ is executed with the same row index _onRowMouseEnter_ is initially executed with.

## Current Behavior
_onRowMouseLeave_ is executed with the row index of the row that replaces the initial row for which _onRowMouseEnter_ was executed.

## Possible Solution
In _FixedDataTableRow_ add the _componentWillUpdate_ React method in order to detect when the current row is being updated with different values (e.g. different row index). This way we can store the original row index before it is replaced by the new one and execute the _onRowMouseLeave_ with the stored row index.

## Steps to Reproduce (for bugs)
1. Hover on a row.
2. Without moving the mouse, scroll down or up. Scroll in such a way that the initial row you hovered on is recycled and replaced in the DOM by another one.
3. The _onRowMouseLeave_ is executed with a different index the _onRowMouseEnter_ was executed with.

Jsfiddle [here](https://jsfiddle.net/sirgandal/obk5h9k3/7/). Relevant information are logged in the console.

Here's a screenshot:
![img1](https://i.imgur.com/4JRxfc0.png)
And another one detailing the error placing some console logs in FixedDataTableRow > ComponentWillUpdate:
![img2](https://i.imgur.com/qRq0PSb.png)

## Context
Our simplified scenario: rows on the table are linked to locations on a map. _onRowMouseEnter_ we want to show a pin on the map and on _onRowMouseLeave_ we want to explicitly remove it. At any single point in time we only want one pin on the map. If _onRowMouseLeave_ is not fired with the correct row index we can end up having multiple pins on the map. Right now we work around this with some additional logic in the functions we pass to _onRowMouseEnter_ and _onRowMouseLeave_ but I think it's better if this is solved in this codebase.   

## Your Environment
* Version used: 0.8.3
* Browser Name and version: Chrome and IE11.
* Operating System and version (desktop or mobile): Windows 10 and Mac OS X.

**Comment by wcjordan** [comment]
Seems reasonable to fix, but looks like your solution will trigger the mouse leave when leaving the row manually after scroll, not when leaving as part of the scroll.  Is that what's desired?

I'm thinking of the case where you:
- Mouse over row 2
- onMouseEnter fires for 2
- Scroll so you're now moused over 22
- No onMouseEnter or onMouseLeave event is fired, you're still showing data for 2, not 22
- Mouse out of 22
- onMouseLeave 2 fires

Wouldn't it be better if scrolling caused onMouseLeave for 2 & onMouseEnter for 22?

**Comment by SirGandal** [comment]
That was the desired behavior since this is already the behavior in most browsers for standard DOM components. For some reason scroll events prevent the mouseleave from being fired properly. See [here](http://jsfiddle.net/tonyleeper/5vwf65f7/).

Not sure about your last question.
If I follow your steps, after the last point I see that it fires onMouseLeave on 2, onMouseEnter on 22 and onMouseLeave on 22. In that order. That's because I don't think it is possible for you to just mouse out of 22 without triggering the on mouse enter on 22 first.

**Comment by wcjordan** [comment]
When I try your fiddle out (in chrome), I see the background update based on my wheel scroll w/o moving my mouse.
![oct 23 2017 5-57 pm](https://user-images.githubusercontent.com/1034455/31915265-cffed3fa-b81b-11e7-91f3-b33517252c0a.gif)

But in the example, no handlers are called unless I actually mouse out.  That's the issue I was thinking off.

**Comment by SirGandal** [comment]
I believe that in the example the mouse leave is fired when you actually stop scrolling. I tried to scroll quite slowly and the mouse leave is called only the moment I stop scrolling. What I would expect from a proper behavior is that the background turns white the exact moment the mouse leaves the blue box because of the scrolling.
![gif](https://i.imgur.com/93lnds9.gif)

**Comment by wcjordan** [comment]
Ya, I think that's correct.  It's still different than what you have implemented, right?

**Comment by KamranAsif** [comment]
+1 to @wcjordan 

I think to achieve the immediate change; instead of firing in onMouseLeave; you should fire in componentsWillRecieveProps.

e.g: onMouseEnter; store data that this row is currenty entered up with rowIndex X
onProp change: If row index has changed and row is currently entered, fire a change and reset those two vars.

This leaves onMouseLeave untouched.

**Comment by wcjordan** [comment]
That sounds like a good idea, and you could probably implement it with a minor improvement on your original solution before the CR suggestions.  Sorry for the run-around.

**Comment by wcjordan** [comment]
We may also want to check the isScrolling prop and skip while scrolling, to avoid perf issues.

**Comment by SirGandal** [comment]
Based on some initial testing it looks like in this case _componentWillReceiveProps_ or _componentWillUpdate_ would serve the same purpose like my original solution.
The approach mentioned by @KamranAsif would still not fire on immediate change because the row is still in the DOM and it will be updated/receive new props only once you scroll far enough. 
Additionally we would need to call the onMouseLeave function with an event as input parameter but we don't really have that event. Am I missing something?

**Comment by KamranAsif** [comment]
Hmm I see. This looks like its because onScroll will block mouseleave requests as well.
The solution becomes a lot more complex then. Basically you would need to know when the mouse is on an element, on X/Y position change if the mouse left it or not. That might lead a to a lot of jank :/

**Comment by SirGandal** [comment]
Yes, I agree.
I wonder if managing this information from the FixedDataTableBufferedRows might help us a bit more. There seems to be more data there that could make that happen.

**Comment by wcjordan** [comment]
Released the base of this with 0.8.4.

I'm going to close, but if we want to follow through on the broader fix (potentially in FDTBufferedRows), feel free to open another ticket focused on that.

Thanks for the PR!
